### ISSUE TITLE
implement CTC with keras?

### ISSUE THREAD

**Issue by blackyang** [fchollet/keras#383]
Hi there,

Has anyone implemented a (Connectionist-Temporal-Classification)CTC loss with keras? 

I attempt to add such a cost function in objectives.py file, based on [rakeshvar's code](https://github.com/rakeshvar/rnn_ctc). The model could be compiled, however, there are several errors when I do model.fit(). I am new to theano so it's really tough for me to debug...

It shouldn't be hard in theory, so I guess I made some "naive" mistakes...

**Comment by fchollet** [comment]
Do you have a reference for what you are trying to implement? As well as your attempt so far.

**Comment by blackyang** [comment]
Hi @fchollet , the original paper of CTC could be found [here](http://www.machinelearning.org/proceedings/icml2006/047_Connectionist_Tempor.pdf) by Alex Graves. 

Basically, CTC is a special loss function to handle alignment. For example, in speech recognition, suppose the input sequence has a length of t (then the output of RNN also has a length of t), usually the target sequence would have a length of w smaller than t. CTC saves the need for pre-segmentation of the inputs and post-segmentation of the net outputs.

I was trying to add a new cost function in **objectives.py**, based on [this ctc.py file](https://github.com/rakeshvar/rnn_ctc). The model could be compiled, however, there are several errors when I do model.fit(). I guess the reason lies in [these lines](https://github.com/fchollet/keras/blob/master/keras/models.py#L249-255), which implies that the two arguments to cost function should share same shape. Correct me if I misunderstand anything

**Comment by futurely** [comment]
@amaas implemented [the CTC loss](https://github.com/amaas/stanford-ctc/blob/master/ctc_fast/ctc-loss/ctc_fast.pyx) strictly faithful to the original paper in a very straightforward way.

**Comment by blackyang** [comment]
@futurely thanks! Currently I am using [this](https://github.com/skaae/Lasagne-CTC) with lasagne :-)

**Comment by amaas** [comment]
It should be relatively straightforward to port our CTC implementation into the Keras framework. Note that our fast version is cython (which doesn't seem to be used elsewhere in Keras). Without cython the loops to compute alignments required to evaluate the CTC loss were painfully slow.

**Comment by jedi00** [comment]
@amaas : do you have a theano version implementation? Or can our fast version work with theano?

**Comment by amaas** [comment]
@jedi00 No, we wrote our RNNs from scratch without Theano. If you want to replace the NN architecture though you could take just our CTC loss and make it a Theano function. It only needs to interact with the final layer so it should be mostly unchanged in a Theano implementation.

**Comment by jinserk** [comment]
Hi @blackyang, did you implement [Lasagne's CTC](https://github.com/skaae/Lasagne-CTC) into Keras? If you did, could you tell me how to do? 
Keras' loss objects are all functions defined in `objectives.py`, and this seems being called from `compile()` function in `models.py`. It is wrapped with `weighted_objective()` function, which call the loss function object with only two params `y_true` and `y_pred`. However, [Lasagne's CTC](https://github.com/skaae/Lasagne-CTC) seems to require 4 params. I'm stuck here. 
Thank you.

**Comment by blackyang** [comment]
Hi @jinserk , I was stuck at the same place, so I used Lasagne which I think is more extensible. By the way I recommend amaas's implementation instead of [Lasagne's CTC](https://github.com/skaae/Lasagne-CTC), since the later one is somehow problematic

**Comment by Michlong** [comment]
I tried it too, unfortunately, failed...

**Comment by futurely** [comment]
The following paper trained a convolutional bidirectional LSTM network to recognize natural scene texts without text line segmentation. The [open source code](http://mc.eistar.net/~xbai/) implemented CTC in C++ for the Torch7 framework in Lua. The C++ code can be modified to use in Python.

[1] B. Shi, X. Bai, C. Yao. [An End-to-End Trainable Neural Network for Image-based Sequence Recognition and Its Application to Scene Text Recognition](http://arxiv.org/pdf/1507.05717v1.pdf). CoRR abs/1507.05717, 2015.

**Comment by ekelsen** [comment]
Baidu just released their open source CPU and GPU implementation of CTC here: 
https://github.com/baidu-research/warp-ctc

It is released as a C-library and bindings for Torch.  The C library should be easy to integrate into many different projects.

**Comment by blackyang** [comment]
@ekelsen thanks for the pointer!

**Comment by ZhangAustin** [comment]
Here is a implementation of Theano bindings for Baidu's warp-ctc: https://github.com/sherjilozair/ctc

Is there any plan for Keras to do bind this?

**Comment by futurely** [comment]
https://github.com/baidu-research/warp-ctc

**Comment by mschonwe** [comment]
And from TensorFlow... https://github.com/tensorflow/tensorflow/tree/master/tensorflow/contrib/ctc

**Comment by shantanudev** [comment]
You guys have any luck with this implementation?

**Comment by daweileng** [comment]
I maintain a repository of CTC with various implementations, including cython, numba/python and theano versions, check here: https://github.com/daweileng/Precise-CTC. You can use CTC_precise or CTC_for_train class, they're both fine for RNN training. 

The CTC objective is different from the current objective functions in Keras, and requires different masking mechanism. I also maintain a repository of Keras MOD with CTC incorporated. Currently, only train_on_batch() is modified to be compatible with CTC. This is enough for me, so there's no definite planning to modify other parts of Keras.

**Comment by shantanudev** [comment]
Oh this is perfect and exactly what I am interested in. Thank you!

**Comment by nouiz** [comment]
Just to let you know, there is this discussion with version that wrap baidu
version that could be faster:

https://github.com/Theano/Theano/issues/3871#issuecomment-207536539

There is 2 current wrapper version at:

https://github.com/mcf06/theano_ctc

and

https://github.com/sherjilozair/ctc

**Comment by lingz** [comment]
@daweileng Do you have any instructions as to how to use your Keras MOD?

**Comment by daweileng** [comment]
Under the repository https://github.com/daweileng/Precise-CTC, there is a folder named as 'Test', you can find a demo script 'mnist_ctc_v4.py' there.

**Comment by vkatsouros** [comment]
In mnist_ctc_v4.py you import from NN_auxiliary and from mytheano_utils. Can you share these too? in Keras MOD?

**Comment by daweileng** [comment]
For who's interested: I updated my CTC-integrated Keras fork to base version 1.0.4, check here: https://github.com/daweileng/keras_MOD/tree/MOD_1.0.4. Till now, the following train/test functions work well with CTC cost:
* train_on_batch()
* test_on_batch()
* predict_on_batch()

**Comment by githubnemo** [comment]
@daweileng Sadly you did not fork the Keras repository. Instead you just copied the files over and added everything (including your patches) in one commit. Can you do that properly (e.g., press the fork button on github, clone, add your changes, commit separately, push) so your patches become actually visible? That'd be awesome.

**Comment by daweileng** [comment]
@githubnemo As explained in the README, the reason I didn't make a pull request is that to avoid a mass modification of Keras' masking mechanism, currently I override **sample_weights** and **masks** variables of Keras. In theory this should not cause problem for other networks but I'm not 100% sure about this. Besides, the modification of **fit()** function is not done yet.

If you just want to know what are changed, you can compare contents of the two repositories.

Progress: Now FCN can work with LSTM + CTC!

**Comment by pasky** [comment]
See also PR#3436.

**Comment by pasky** [comment]
See also #3436

**Comment by harikrishnavydana** [comment]
dear @fchollet  the ocr example dorent run propserly and generates multiple error using the generator? can u please check it again

**Comment by patyork** [comment]
@harikrishnavydana The [ocr ](https://raw.githubusercontent.com/fchollet/keras/master/examples/image_ocr.py)example runs fine for me on both Theano and Tensorflow.

If it is not working for you, please review the issue guidelines (update keras) and if the issue persists, open a new issue.

**Comment by harikrishnavydana** [comment]
Thank you, i was using the older version of keras @patyork

**Comment by besanson** [comment]
Hi, thanks @patyork . Just to understand. You are putting text in images. And using some of these to train and others to validate? but you are using full words to train and not characters. Pycairo is a complicated library to install :)

**Comment by anuj-rathore** [comment]
I am trying to use keras ctc in Bidirectional LSTM i.e. https://github.com/lvapeab/ABiViRNet
Network is as:
https://pastebin.com/9QXbJSwE

When I am not able to integrate  keras batch_ctc into it

**Comment by selcouthlyBlue** [comment]
Apparently, there is a ctc_loss implementation in Keras. There's an open [issue](https://github.com/keras-team/keras/issues/7049) on Keras' `ctc_batch_cost` in the tensorflow_backend.

**Comment by hypernote** [comment]
Hello... We already have some sample of CTC at keras repository?

**Comment by selcouthlyBlue** [comment]
You mean this [one](https://github.com/keras-team/keras/blob/master/examples/image_ocr.py)? If so, yeah I know there is already a sample. It's just when I search for "Keras CTC" in google, this issue comes up and I just thought it would be nice to let people know that such an implementation already exists.

**Comment by hypernote** [comment]
Great

**Comment by rasto2211** [comment]
Is it ok to use `ctc_batch_cost` as keras loss function and pass it to `model.compile`? All the losses that are implemented in:
https://github.com/keras-team/keras/blob/master/keras/losses.py
take only one sample. Is it efficient?

Is there any plan to integrate [WarpCTC](https://github.com/baidu-research/warp-ctc) to Keras?

**Comment by saisumanth007** [comment]
Could you please tell what input_length and label_length specify? As per the documentation it seems label_length contains the lengths of ground truth strings (in case of OCR). But I'm not sure what input_length means.

**Comment by aayushee** [comment]
I think input_length refers to your sequence length and label_length refers to the ground truth label length.
